Add Premultiplied Alpha tutorial (#2531)
pygame-community/pygame-ce: docs/reST/tutorials/en/premultiplied-alpha.rst

.. TUTORIAL:What is Premultiplied Alpha?

.. include:: ../../common.txt

****************************************************
  Pygame Tutorials - What is Premultiplied Alpha?
****************************************************

What is Premultiplied Alpha?
============================

.. rst-class:: docinfo

:Author: Dan Lawrence

> :Contact: danintheshed@gmail.com

Author: Not sure if I should be slapping my email in here - I just copied the format from the other tutorials.

Reviewer: its your choice I guess.

Author: eh, I'll leave it.

Introduction to Alpha Composition
---------------------------------

Alpha composition is the process by which we combine one or more semi-transparent images
into a final non-transparent image for display. In pygame-ce these images are called Surfaces
so I'll use that terminology going forwards.

The simplest example of an alpha composition is when we create a Surface, fill it with a color
that has an alpha channel, and then blit that surface direct to our display surface. For example:

.. code-block:: python
   :caption: Simple example of Straight Alpha composition
   :name: straight_alpha.py
   :linenos:

    import pygame
    from pygame import SRCALPHA, QUIT

    pygame.init()

    pygame.display.set_caption("Basic Composition")
    display_surf = pygame.display.set_mode((300, 170))

>     basic_surf = pygame.Surface((120, 120), flags=SRCALPHA)

Reviewer: maybe explain the flag `SRCALPHA` and why it is needed.

I would suggest to add a comment for the `25` of the fill color. Because this is a tutorial.

Author: Done.

want to use it over straight alpha?

There are two main reasons - and the first is performance. As you saw in the two formulas above, there
is one less mathematical operation to do at composition time with premultiplied alpha. Assuming you are
not adjusting your alpha in real-time - and in most game development usages you won't be, that is one
less operation to do per pixel which means, that on average your premultiplied alpha blits will be a
little bit faster than your straight alpha blits.

The second reason is a little more complicated to demonstrate so I've prepared a program to demonstrate
the issue. Essentially the straight alpha formula has issues when blending together two surfaces that
both contain alpha pixels, and the impact of this can vary from not noticeable at all to looking very
messy. Here is the example program for straight alpha:

.. code-block:: python
   :caption: Example of Straight Alpha composition between two surfaces with per-pixel alpha
   :name: straight_alpha_tool_tip.py
   :linenos:

    import pygame

    pygame.init()

    pygame.display.set_caption("Straight Alpha")
    display_surf = pygame.display.set_mode((300, 170))

>     text_font = pygame.font.Font("fonts/verdana.ttf", size=12)

Reviewer: why not use the default font for simplicity?

What is Premultiplied Alpha?
============================

.. rst-class:: docinfo

:Author: Dan Lawrence
:Contact: danintheshed@gmail.com

Introduction to Alpha Composition
---------------------------------

Alpha composition is the process by which we combine one or more semi-transparent images
into a final non-transparent image for display. In pygame-ce these images are called Surfaces
so I'll use that terminology going forwards.

The simplest example of an alpha composition is when we create a Surface, fill it with a color
that has an alpha channel, and then blit that surface direct to our display surface. For example:

.. code-block:: python
   :caption: Simple example of Straight Alpha composition
   :name: straight_alpha.py
   :linenos:

    import pygame
    from pygame import SRCALPHA, QUIT

    pygame.init()

    pygame.display.set_caption("Basic Composition")
    display_surf = pygame.display.set_mode((300, 170))

    # create a Surface with the SRCALPHA flag to add an extra channel of data
    # to each pixel that indicates how transparent it should be (from 0 - fully
    # transparent, to 255 - fully opaque.
    basic_surf = pygame.Surface((120, 120), flags=SRCALPHA)

>     # the fourth number here sets the alpha to 25 (highly translucent)

Reviewer 2: "highly translucent" this is confusing wording (atleast to me, as a non-native speaker). I get what translucent means, but saying "highly", it is not clear which way that highly goes. Maybe it would be better say "almost but not fully transparent" as this leaves no room for confusion. However you can resolve this any way you see fit, or even keep it this way.

Reviewer 3: "almost fully see through?"

>     # the fourth number here sets the alpha to 25 (highly translucent)

Author: ```suggestion
    # the fourth number here sets the alpha to 25 (almost fully see through)
```

>     basic_surf.fill(pygame.Color(50, 50, 50, 25))

Author: ```suggestion
    basic_surf.fill((50, 50, 50, 25))
```

another and the simplicity of its use. There is no setup required and most image editing programs will
export alpha in this way as a separate channel. This is why it is the default composition method in
pygame-ce.

How to use Premultiplied Alpha blending
---------------------------------------

Premultiplied alpha blending uses a slightly different formula to compose two surfaces ::

    result = source.RGB + (destination.RGB * (1 - source.A))

As you can see there is one less multiplication in there - this is because, as is implied by the name,
in a premultiplied alpha composition all the pixels colors have already been multiplied by their alpha
channel value.

We can rewrite the example above to use premultiplied alpha:

.. code-block:: python
   :caption: Simple example of Premultiplied Alpha composition
   :name: premultiplied_alpha.py
   :linenos:
   :emphasize-lines: 11,23

    import pygame
    from pygame import SRCALPHA, BLEND_PREMULTIPLIED, QUIT

    pygame.init()

    pygame.display.set_caption("Basic Composition")
    display_surf = pygame.display.set_mode((300, 170))

    basic_surf = pygame.Surface((120, 120), flags=SRCALPHA)

>     basic_surf.fill(pygame.Color(50, 50, 50, 25))

Author: ```suggestion
    basic_surf.fill((50, 50, 50, 25))
```

not adjusting your alpha in real-time - and in most game development usages you won't be, that is one
less operation to do per pixel which means, that on average your premultiplied alpha blits will be a
little bit faster than your straight alpha blits.

The second reason is a little more complicated to demonstrate so I've prepared a program to demonstrate
the issue. Essentially the straight alpha formula has issues when blending together two surfaces that
both contain alpha pixels, and the impact of this can vary from not noticeable at all to looking very
messy. Here is the example program for straight alpha:

.. code-block:: python
   :caption: Example of Straight Alpha composition between two surfaces with per-pixel alpha
   :name: straight_alpha_tool_tip.py
   :linenos:

    import pygame

    pygame.init()

    pygame.display.set_caption("Straight Alpha")
    display_surf = pygame.display.set_mode((300, 170))

    text_font = pygame.font.Font("fonts/verdana.ttf", size=12)

    tool_tip_text = text_font.render(
        "Some text in a box, to test alpha blending. "
        "A quick brown fox jumps over the lazy dog.",
        True,

>         pygame.Color((200, 200, 250)),

Author: ```suggestion
        (200, 200, 250),
```

        wraplength=100,
    ).convert_alpha()

    tool_tip_surf = pygame.Surface((120, 120), flags=pygame.SRCALPHA)

>     tool_tip_surf.fill(pygame.Color(50, 50, 50, 25))

Author: ```suggestion
    tool_tip_surf.fill((50, 50, 50, 25))
```

copy it into a 'fonts/' subdirectory or use an alternative font. The issue is visible on all
fonts but more obvious on some fonts than others depending on how much they rely on alpha pixels
for visibility. If you run this program you will get a result that looks like this:

.. image:: ../assets/straight_alpha_composition.png
   :alt: Example of Straight Alpha Composition

Which, to my eyes, makes the text difficult to read and something of a strain on the eyes.

If we rewrite the example to use premultiplied alpha composition instead:

.. code-block:: python
   :caption: Example of Premultiplied Alpha composition between two surfaces with per-pixel alpha
   :name: premultiplied_alpha_tool_tip.py
   :linenos:
   :emphasize-lines: 5,17,21,24,36

    import pygame

    pygame.init()

    pygame.display.set_caption("Premultiplied Alpha")
    display_surf = pygame.display.set_mode((300, 170))

    text_font = pygame.font.Font("fonts/verdana.ttf", size=12)

    tool_tip_text = text_font.render(
        "Some text in a box, to test alpha blending. "
        "A quick brown fox jumps over the lazy dog.",
        True,

>         pygame.Color((200, 200, 250)),

Author: ```suggestion
        (200, 200, 250),
```

        wraplength=100,
    ).convert_alpha()
    tool_tip_text = tool_tip_text.premul_alpha()

    tool_tip_surf = pygame.Surface((120, 120), flags=pygame.SRCALPHA)

>     tool_tip_surf.fill(pygame.Color(50, 50, 50, 25))

Author: ```suggestion
    tool_tip_surf.fill((50, 50, 50, 25))
```
